File: CartMinder/app/src/main/java/id/ac/ui/cs/mobileprogramming/kevinlh/cartminder/model/Item.java
package id.ac.ui.cs.mobileprogramming.kevinlh.cartminder.model;

import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.PrimaryKey;

import org.jetbrains.annotations.NotNull;

import java.util.Locale;


@Entity(tableName = "item_table")
public class Item {
    @PrimaryKey(autoGenerate = true)
    private long id;
    @ForeignKey(
            entity = Cart.class,
            parentColumns = "id",
            childColumns = "cartId",
            onDelete = ForeignKey.CASCADE
    )
    private long cartId;
    @ForeignKey(
            entity = ItemDetail.class,
            parentColumns = "id",
            childColumns = "detailId",
            onDelete = ForeignKey.NO_ACTION
    )
    private long detailId;
    private String title;
    private String description;
    private int price;

    public Item() {
        this("", "", 0);
    }

    public Item(String title, String description, int price) {
        this.id = 0;
        this.title = title;
        this.description = description;
        this.price = price;
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public long getCartId() {
        return cartId;
    }

    public void setCartId(long cartId) {
        this.cartId = cartId;
    }

    public long getDetailId() {
        return detailId;
    }

    public void setDetailId(long detailId) {
        this.detailId = detailId;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public int getPrice() {
        return price;
    }

    public void setPrice(int price) {
        this.price = price;
    }

    public String getPriceToString() {
        return String.format(Locale.US, "%,d IDR", price);
    }

    @NotNull
    @Override
    public String toString() {
        return "Item{" +
                "id=" + id +
                ", cartId=" + cartId +
                ", detailId=" + detailId +
                ", title='" + title + '\'' +
                ", description='" + description + '\'' +
                ", price=" + price +
                '}';
    }
}

File: CartMinder2/app/src/main/java/id/ac/ui/cs/mobileprogramming/kevinlh/cartminder2/model/Item.kt
package id.ac.ui.cs.mobileprogramming.kevinlh.cartminder2.model

import android.annotation.SuppressLint
import android.os.Parcel
import android.os.Parcelable
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.PrimaryKey

@Entity(tableName = "item_table")
class Item() : Parcelable {
    @PrimaryKey(autoGenerate = true)
    var id: Long = 0

    @ForeignKey(
        entity = Cart::class,
        parentColumns = ["id"],
        childColumns = ["cartId"],
        onDelete = ForeignKey.CASCADE
    )
    var cartId: Long = 0
    var title: String = ""
    var price: Int = 0

    override fun equals(other: Any?): Boolean {
        if (other is Item)
            return cartId == other.cartId &&
                    title == other.title &&
                    price == other.price
        throw RuntimeException("can't compare to non Item object")
    }

    @SuppressLint("DefaultLocale")
    override fun toString(): String {
        val s = StringBuilder().apply {
            append("Item::\n")
            append("  title: $title\n")
            append("  cartId: $cartId\n")
            append("  price: $price\n")
        }
        return s.toString()
    }

    constructor(parcel: Parcel) : this() {
        id = parcel.readLong()
        cartId = parcel.readLong()
        title = parcel.readString() ?: ""
        price = parcel.readInt()
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeLong(id)
        parcel.writeLong(cartId)
        parcel.writeString(title)
        parcel.writeInt(price)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<Item> {
        override fun createFromParcel(parcel: Parcel): Item {
            return Item(parcel)
        }

        override fun newArray(size: Int): Array<Item?> {
            return arrayOfNulls(size)
        }
    }
}